Subject: Udhcpc: when decoding DHCP_SUBNET, ensure it is 4 bytes long
ID: CVE-2018-20679
Origin: 1_30_0-35-g74d9f1ba3 <https://git.busybox.net/busybox/commit/?id=1_30_0-35-g74d9f1ba3>
Upstream-Author: Denys Vlasenko <vda.linux@googlemail.com>
Date: Mon Jan 7 15:33:42 2019 +0100
Bug-Debian: https://bugs.debian.org/918846

    function                                             old     new   delta
    udhcp_run_script                                     795     801      +6
    
    Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>

--- a/networking/udhcp/common.c
+++ b/networking/udhcp/common.c
@@ -302,7 +302,7 @@
 {
 	uint8_t *r = udhcp_get_option(packet, code);
 	if (r) {
-		if (r[-1] != 4)
+		if (r[-OPT_DATA + OPT_LEN] != 4)
 			r = NULL;
 	}
 	return r;
--- a/networking/udhcp/common.h
+++ b/networking/udhcp/common.h
@@ -119,7 +119,7 @@
 //#define DHCP_TIME_SERVER      0x04 /* RFC 868 time server (32-bit, 0 = 1.1.1900) */
 //#define DHCP_NAME_SERVER      0x05 /* IEN 116 _really_ ancient kind of NS */
 //#define DHCP_DNS_SERVER       0x06
-//#define DHCP_LOG_SERVER       0x07 /* port 704 UDP log (not syslog)
+//#define DHCP_LOG_SERVER       0x07 /* port 704 UDP log (not syslog) */
 //#define DHCP_COOKIE_SERVER    0x08 /* "quote of the day" server */
 //#define DHCP_LPR_SERVER       0x09
 #define DHCP_HOST_NAME          0x0c /* 12: either client informs server or server gives name to client */
--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -531,7 +531,7 @@
 		temp = udhcp_get_option(packet, code);
 		*curr = xmalloc_optname_optval(temp, &dhcp_optflags[i], opt_name);
 		putenv(*curr++);
-		if (code == DHCP_SUBNET) {
+		if (code == DHCP_SUBNET && temp[-OPT_DATA + OPT_LEN] == 4) {
 			/* Subnet option: make things like "$ip/$mask" possible */
 			uint32_t subnet;
 			move_from_unaligned32(subnet, temp);
